### PR TITLE
MariaDB 10.2.7+

### PR DESCRIPTION
**Pull request by VicDeo**
## Description
Backup solution for #28695 
I'm not happy with this solution but if upstream doesn't manage to fix the issue until 10.0.4 we'll need it

Here we bypass native Doctrine/DBAL   `MySqlSchemaManager::_getPortableTableColumnDefinition` implementation for  MariaDB >= 10.2.7 and use customized implementation that unquotes default values.
All DB engines from non-MySQL family work as before. For MySQL family a new listener is registered.
Triggering native/customized codeflow is done inside the listener itself.

## Related Issue
 #28695  

## Motivation and Context
Fixes MariaDB 10.2.7+ support

## How Has This Been Tested?
By installing ownCloud with  MariaDB 10.2.7+ as a DB engine

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by PVince81** [comment]
@VicDeo I'm setting this to p1 to keep it in view before releasing 10.0.4. When the time comes we'll decide whether to use the PR or upstream depending on the progress.

**Comment by VicDeo** [comment]
Stable10: https://github.com/owncloud/core/pull/29240

**Comment by PVince81** [comment]
It looks like it's likely that we won't be able to use the fix from upstream because it will likely not be backported: https://github.com/doctrine/dbal/pull/2825#issuecomment-336413059

And OC 10 still needs to support PHP 5.6.

So I suggest we merge this for now and kill it once we deprecate PHP 7.0...

@DeepDiver1975 do you still have objections ?

**Comment by PVince81** [comment]
let's get an explicit statement here: https://github.com/doctrine/dbal/pull/2825#issuecomment-336821026

**Comment by PVince81** [comment]
seems no upstream backport https://github.com/doctrine/dbal/pull/2825#issuecomment-340414478

**Comment by PVince81** [comment]
@VicDeo please write a unit test for the listener and make codecov happy.

If we're going to carry this workaround long term, better have it properly covered.
Thanks.

**Comment by VicDeo** [comment]
@PVince81 oook... :\
 (I already checked doctrine unit tests and know that they are not reusable here...)

**Comment by DeepDiver1975** [comment]
> If we're going to carry this workaround long term, better have it properly covered.
> Thanks.

@SergioBertolinSG is working on getting coverage for mysql specific code - see #29410 
Once merged I assume codecov will be happy here as well ...

**Comment by DeepDiver1975** [comment]
> @SergioBertolinSG is working on getting coverage for mysql specific code - see #29410
> Once merged I assume codecov will be happy here as well ...

looks like we need to add mariadb 10.2.7 to the test matrix

**Comment by PVince81** [comment]
can we get this finished for RC1?

**Comment by DeepDiver1975** [comment]
> can we get this finished for RC1?
> Merge state

I quickly added coverage report for mariadb - lets see ....

**Comment by DeepDiver1975** [comment]
I guess mariadb hates us ... first errors popping up ...

````
Runtime:       PHPDBG 7.1.10-1+ubuntu16.04.1+deb.sury.org+1
Configuration: /drone/src/tests/phpunit-autotest.xml

.............................................................   61 / 8641 (  0%)
........................S....................................  122 / 8641 (  1%)
.............................................................  183 / 8641 (  2%)
.............................................................  244 / 8641 (  2%)
.............................................................  305 / 8641 (  3%)
.............................................................  366 / 8641 (  4%)
.............................................................  427 / 8641 (  4%)
.............................................................  488 / 8641 (  5%)
.............................................................  549 / 8641 (  6%)
.............................................................  610 / 8641 (  7%)
.............................................................  671 / 8641 (  7%)
.............................................................  732 / 8641 (  8%)
.............................................................  793 / 8641 (  9%)
.............................................................  854 / 8641 (  9%)
.............................................................  915 / 8641 ( 10%)
.............................................................  976 / 8641 ( 11%)
................S............................................ 1037 / 8641 ( 12%)
............................................................. 1098 / 8641 ( 12%)
............................................................. 1159 / 8641 ( 13%)
.........................................................FFFS 1220 / 8641 ( 14%)
S............................................................ 1281 / 8641 ( 14%)
............................................................. 1342 / 8641 ( 15%)
............................................................. 1403 / 8641 ( 16%)
..................E...................................E...... 1464 / 8641 ( 16%)
............................................................. 1525 / 8641 ( 17%)
......................E...................................... 1586 / 8641 ( 18%)
....E........................................................ 1647 / 8641 ( 19%)
............................................................. 1708 / 8641 ( 19%)
............................................................. 1769 / 8641 ( 20%)
............................................................. 1830 / 8641 ( 21%)
............................................................. 1891 / 8641 ( 21%)

**Comment by DeepDiver1975** [comment]
@VicDeo can you please have a look at the failing tests? looks really fishy

**Comment by VicDeo** [comment]
@DeepDiver1975 yes, looking right now.
`Doctrine\DBAL\Driver\PDOException: SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect string value: '\xF0\x9F\x98\x90, ...' for column 'path' at row 1` seems to related to utf8mb4

Whil `Test\DB\SchemaDiffTest::testZeroChangeOnSchemaMigrations`  failures are  fishy indeed.

**Comment by VicDeo** [comment]
@DeepDiver1975 @PVince81 Tests related to false positives with default values are passing now.
Can't reproduce  failures related to 4 byte utf-8 with my MariaDB.

**Comment by VicDeo** [comment]
Ok. Drone uses MariaDb 10.3.2 
Variable 'innodb_file_format' doesn't exist in MariaDB  since 10.3 
See https://mariadb.com/kb/en/library/changes-improvements-in-mariadb-103/

`innodb_file_format` is used for utf8mb4 detection
https://github.com/owncloud/core/blob/master/lib/private/DB/MySqlTools.php#L36

**Comment by PVince81** [comment]
@VicDeo if mb4 is default in 10.3, we might need some kind of migration for the data... raise new ticket ?

**Comment by PVince81** [comment]
@VicDeo I've raised https://github.com/owncloud/core/issues/29483 to look into the absence of said switch.

As @DeepDiver1975 said we need to define max versions for databases. This means we might need to fix our Drone setup for now to test the older MariaDB and bypass the missing innodb switch. @SergioBertolinSG

**Comment by VicDeo** [comment]
@PVince81 as you see everything passes here. I adjusted MariaDB version in the very last commit

File: lib/private/DB/MySqlSchemaColumnDefinitionListener.php [hunk 1]
```diff
@@ -51,10 +53,18 @@ public function onSchemaColumnDefinition(SchemaColumnDefinitionEventArgs $eventA
 			$this->_platform = \OC::$server->getDatabaseConnection()->getDatabasePlatform();
 		}
 
-		$tableColumn = $eventArgs->getTableColumn();
-		$eventArgs->preventDefault();
-		$column = $this->_getPortableTableColumnDefinition($tableColumn);
-		$eventArgs->setColumn($column);
+		$version = \OC::$server->getDatabaseConnection()->getDatabaseVersionString();
+		$mariadb = false !== stripos($version, 'mariadb');
+		if ($mariadb && version_compare($this->getMariaDbMysqlVersionNumber($version), '10.2.7', '>=')) {
```
Comment (PVince81) [inline lib/private/DB/MySqlSchemaColumnDefinitionListener.php]: 10.2.7, really ? not even 10.2.0 ? 😢

File: lib/private/DB/MySqlSchemaColumnDefinitionListener.php [hunk 2]
```diff
@@ -51,10 +53,18 @@ public function onSchemaColumnDefinition(SchemaColumnDefinitionEventArgs $eventA
 			$this->_platform = \OC::$server->getDatabaseConnection()->getDatabasePlatform();
 		}
 
-		$tableColumn = $eventArgs->getTableColumn();
-		$eventArgs->preventDefault();
-		$column = $this->_getPortableTableColumnDefinition($tableColumn);
-		$eventArgs->setColumn($column);
+		$version = \OC::$server->getDatabaseConnection()->getDatabaseVersionString();
+		$mariadb = false !== stripos($version, 'mariadb');
```
Comment (PVince81) [inline lib/private/DB/MySqlSchemaColumnDefinitionListener.php]: what if $version is null ?

**Review by PVince81 — COMMENTED** [review]
Code looks good to me. See comments.

If @DeepDiver1975 also agrees with the change then let's merge.

File: lib/private/DB/MySqlSchemaColumnDefinitionListener.php (hunk 2 above)
Comment (VicDeo) [reply lib/private/DB/MySqlSchemaColumnDefinitionListener.php]: @PVince81 `stripos` will return false without a warning

File: lib/private/DB/MySqlSchemaColumnDefinitionListener.php (hunk 1 above)
Comment (VicDeo) [reply lib/private/DB/MySqlSchemaColumnDefinitionListener.php]: @PVince81 https://jira.mariadb.org/browse/MDEV-13132
Fix Version/s: 10.2.7

**Review by PVince81 — APPROVED** [review]
:+1:

**Pull request merged**